packages/dynamic_layouts/lib/src/render_dynamic_grid.dart

```diff
@@ -0,0 +1,410 @@
+  /// position are determined by a delegate.
+  ///
+  /// The [childManager] and [gridDelegate] arguments must not be null.
+  RenderDynamicSliverGrid({
+    required super.childManager,
+    required SliverGridDelegate gridDelegate,
+  }) : assert(gridDelegate != null),
+       _gridDelegate = gridDelegate;
+
+  @override
+  void setupParentData(RenderObject child) {
+    if (child.parentData is! SliverGridParentData) {
+      child.parentData = SliverGridParentData();
+    }
+  }
+
+  /// The delegate that controls the size and position of the children.
+  SliverGridDelegate get gridDelegate => _gridDelegate;
+  SliverGridDelegate _gridDelegate;
+  set gridDelegate(SliverGridDelegate value) {
+    assert(value != null);
+    if (_gridDelegate == value) {
+      return;
+    }
+    if (value.runtimeType != _gridDelegate.runtimeType ||
+        value.shouldRelayout(_gridDelegate)) {
+      markNeedsLayout();
+    }
+    _gridDelegate = value;
+  }
+
+  @override
+  double childCrossAxisPosition(RenderBox child) {
+    final SliverGridParentData childParentData = child.parentData! as SliverGridParentData;
+    return childParentData.crossAxisOffset!;
+  }
+
+  @override
+  void performLayout() {
+    final SliverConstraints constraints = this.constraints;
+    childManager.didStartLayout();
+    childManager.setDidUnderflow(false);
+
+    final double scrollOffset = constraints.scrollOffset + constraints.cacheOrigin;
+    assert(scrollOffset >= 0.0);
+    final double remainingExtent = constraints.remainingCacheExtent;
+    assert(remainingExtent >= 0.0);
+    final double targetEndScrollOffset = scrollOffset + remainingExtent;
+    final DynamicSliverGridLayout layout = _gridDelegate.getLayout(constraints) as DynamicSliverGridLayout;
+    int leadingGarbage = 0;
+    int trailingGarbage = 0;
+    bool reachedEnd = false;
+
+    // This algorithm in principle is straight-forward: find the first child
+    // that overlaps the given scrollOffset, creating more children at the top
+    // of the grid if necessary, then walk through the grid updating and laying
+    // out each child and adding more at the end if necessary until we have
+    // enough children to cover the entire viewport.
+    //
+    // It is complicated by one minor issue, which is that any time you update
+    // or create a child, it's possible that the some of the children that
+    // haven't yet been laid out will be removed, leaving the list in an
+    // inconsistent state, and requiring that missing nodes be recreated.
+    //
+    // To keep this mess tractable, this algorithm starts from what is currently
+    // the first child, if any, and then walks up and/or down from there, so
+    // that the nodes that might get removed are always at the edges of what has
+    // already been laid out.
+
+    // Make sure we have at least one child to start from.
+    if (firstChild == null) {
+      if (!addInitialChild()) {
```

Comment:
could be one if?

Reply:
Sure! I think originally I was following the same pattern in RenderSliverList so it is easy to follow and maintain the similar logic between them, maybe I can do a clean up in RenderSliverList later.